tests/test_gym_collector.py

```diff
@@ -0,0 +1,109 @@
+import torch
+
+from gymnasium.vector import AsyncVectorEnv
+from tests.gym import DictGymWrapper, HitTheMiddle, SimpleGymCollector
+from torch import nn
+
+from emote import Trainer
+from emote.callbacks import BatchCallback
+from emote.memory import MemoryLoader, TableMemoryProxy
+from emote.memory.builder import DictObsTable
+from emote.sac import FeatureAgentProxy
+from emote.trainer import TrainingShutdownException
+
+
+class RandomPolicy(nn.Module):
+    def __init__(self, action_dim: int):
+        super().__init__()
+        self.action_dim = action_dim
+
+    def forward(self, obs: torch.Tensor):
+        batch_size = obs.shape[0]
+        rand_actions = 2 * (torch.rand(batch_size, self.action_dim) - 0.5)
+        return rand_actions, 0
+
+
+class HitTheMiddleDataInspector(BatchCallback):
+    def __init__(
+        self,
+        num_bp: int,
+        device: torch.device,
+        data_group: str = "default",
+    ):
+        super().__init__()
+        self.data_group = data_group
+        self.cycle = num_bp
+        self.device = device
+
+    def begin_batch(self, *args, **kwargs):
+        obs, next_obs, action, reward = self.get_batch(*args, **kwargs)
+        batch_size = obs.shape[0]
+        sim_next_obs, sim_reward = self.simulate_hit_the_middle(action, obs)
+        for i in range(batch_size):
+            obs_err = torch.mean(torch.abs(sim_next_obs[i] - next_obs[i])).detach()
+            reward_err = torch.mean(torch.abs(sim_reward[i] - reward[i])).detach()
+            if obs_err > 0.001 or reward_err > 0.001:
+                print(f"obs_err: {obs_err}, reward_err: {reward_err}")
+                print(f"obs: {obs[i]}, action: {action[i]}")
+                print(f"reward: {reward[i]}\nsim_reward: {sim_reward[i]}")
+                print(f"next_obs: {next_obs[i]}\nsim_next_obs: {sim_next_obs[i]}")
+                raise ValueError(
+                    "Loaded values for obs/reward does not match the calculated ones"
+                )
```

Comment:
I'd suggest doing this on the following format:

```
message = f'''
    obs_err ..
    obs ...
    reward ...
'''
raise ValueError("Loaded... {message}")
```

That will make it clearer in pytest I think, in case there's lots of logging etc. In py3.11 this is an *excellent* use-case for PEP678 too.

Reply:
Fixed!